walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -63,6 +79,37 @@
       freeTimes.add(TimeRange.fromStartEnd(startTime, TimeRange.END_OF_DAY, /* inclusive */ true));
     }
 
-    return freeTimes;
+    // Repeat the exact same search, but this time find meetings
+    // that only mandatory attendees can attend.
+    startTime = TimeRange.START_OF_DAY;
+    ArrayList<TimeRange> freeTimesMandatory = new ArrayList<>();
+
+    // If there are restrictions for only mandatory attendees.
+    if (busyTimesMandatory.size() > 0) {
+      // Find gaps between events to schedule the meeting
+      // for mandatory attendees only.
+      for (TimeRange busy : busyTimesMandatory) {
+        if (busy.start() - startTime >= request.getDuration()) {
+          freeTimesMandatory.add(
+              TimeRange.fromStartEnd(startTime, busy.start(), /* inclusive */ false));
+        }
+        if (startTime < busy.end()) {
+          startTime = busy.end();
+        }
+      }
+
+      if (TimeRange.END_OF_DAY - startTime >= request.getDuration()) {
+        freeTimesMandatory.add(
+            TimeRange.fromStartEnd(startTime, TimeRange.END_OF_DAY, /* inclusive */ true));
+      }
+    }
+
+    // If there is at least one time where both mandatory
+    // and optional attendees can attend the meeting, return that list.
+    if (freeTimes.size() > 0) {
+      return freeTimes;
+    }
```

Comment:
You can move this chunk before the Mandatory list search (after line 80). If there are free times that all attendees can attend, don't have to look through the busyTimesMandatory list.